### PR TITLE
Flink: Maintenance - Lock remover

### PR DESCRIPTION
**Pull request by pvary**
Operator to remove the lock at the end of the Maintenance flow.
https://docs.google.com/document/d/16g3vR18mVBy8jbFaLjf2JwAANuYOmIwr15yDDxovdnA/edit#heading=h.lt9eaimi6zyz

### CODE REVIEW

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    if (value.success()) {
+      successfulStreamResultCounterMap.get(value.taskId()).inc();
+    } else {
+      failedStreamResultCounterMap.get(value.taskId()).inc();
+    }
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) {
+    if (mark.getTimestamp() > lastProcessed) {
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:115]: should  this be `>` or `>=`?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        taskNames != null && !taskNames.isEmpty(), "Invalid task names: null or empty");
+
+    this.lockFactory = lockFactory;
+    this.taskNames = taskNames;
+  }
+
+  @Override
+  public void open() throws Exception {
+    super.open();
+    this.successfulStreamResultCounterMap = Maps.newHashMapWithExpectedSize(taskNames.size());
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:63]: Good point!
Remainder from a previous version

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+  @Override
+  public void processElement(StreamRecord<TaskResult> element) {
+    TaskResult value = element.getValue();
+    LOG.debug("TaskResult {} arrived", value);
+    long length = System.currentTimeMillis() - value.startEpoch();
+    output.collect(
+        new StreamRecord<>(
+            new MaintenanceResult(
+                value.startEpoch(), value.taskId(), length, value.success(), value.exceptions())));
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:100]: I'm open to discussion here.
Here are my thoughts:
- I prefer to have an output of the stream which could be logged or processed later. Having `Void` and changing later is backward incompatible. Having something for which we could add attributes could be compatible. That is why I have created the `MaintenanceResult`
- I would like to do the `duration` calculation (and metrics logging) in a single place, so I put it in here instead of putting it in every `Action`
- Inheritance for me means a bit more that some attributes are overlap. If we later need changes in one of the `MaintenanceResult` or the `TaskResult` it is easier to understand that they are independent.

Rejected alternatives:
- Keep only the `MaintenanceResult ` and fill the `duration` later - I hate "not filled" attributes which are later generated, because it is hard to follow when and why they got the value

Your thoughts?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    lock.unlock();
+    this.lastProcessed = value.startEpoch();
+
+    // Update the metrics
+    lastRunLength.get(value.taskId()).set(length);
+    if (value.success()) {
+      successfulStreamResultCounterMap.get(value.taskId()).inc();
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:107]: renamed to `taskIndex`

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    } else {
+      failedStreamResultCounterMap.get(value.taskId()).inc();
+    }
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) {
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:114]: Yeah, this is how we will calculate the watermark

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MaintenanceResult.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    return this.startEpoch == that.startEpoch
+        && this.taskId == that.taskId
+        && this.length == that.length
+        && this.success == that.success
+        && this.exceptions.size() == that.exceptions.size();
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(startEpoch, taskId, length, success, exceptions.size());
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MaintenanceResult.java:91]: Removed `hashCode` and `equals` for now. Used that for testing purposes on my later changes. We can decide if we want them or not there.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResult.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  public boolean success() {
+    return success;
+  }
+
+  public List<Exception> exceptions() {
+    return exceptions;
+  }
+
+  @Override
+  public String toString() {
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResult.java:57]: Neither `MaintenanceResult` for now

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MaintenanceResult.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    } else if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+
+    MaintenanceResult that = (MaintenanceResult) other;
+    return this.startEpoch == that.startEpoch
+        && this.taskId == that.taskId
+        && this.length == that.length
+        && this.success == that.success
+        && this.exceptions.size() == that.exceptions.size();
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MaintenanceResult.java:86]: Removed for now

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  private final List<String> taskNames;
+
+  private transient Map<Integer, Counter> successfulStreamResultCounterMap;
+  private transient Map<Integer, Counter> failedStreamResultCounterMap;
+  private transient Map<Integer, AtomicLong> lastRunLength;
+  private transient TriggerLockFactory.Lock lock;
+  private transient TriggerLockFactory.Lock recoveryLock;
+  private transient long lastProcessed = 0L;
+
+  public LockRemover(TriggerLockFactory lockFactory, List<String> taskNames) {
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:51]: I'm not sure about this.

There were `Actions` already for Flink before this PR. This referred to a single batch job which could be executed. See: https://github.com/apache/iceberg/tree/64b36999d7ff716ae2534fb0972fcc10d22a64c2/flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/actions

I would like to avoid the confusion with that as well.

In the doc/comments before I have called it `Maintenance Task` everywhere.
Shall we rename to `maintenanceTaskNames`? Do you have any better ideas?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: done

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    long length = System.currentTimeMillis() - value.startEpoch();
+    output.collect(
+        new StreamRecord<>(
+            new MaintenanceResult(
+                value.startEpoch(), value.taskId(), length, value.success(), value.exceptions())));
+    lock.unlock();
+    this.lastProcessed = value.startEpoch();
+
+    // Update the metrics
+    lastRunLength.get(value.taskId()).set(length);
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Done

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestLockRemover.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    source
+        .dataStream()
+        .transform(
+            DUMMY_NAME,
+            TypeInformation.of(MaintenanceResult.class),
+            new LockRemover(new TestingLockFactory(), Lists.newArrayList(TASK_0, TASK_1)))
+        .setParallelism(1)
+        .sinkTo(result);
+
+    JobClient jobClient = null;
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestLockRemover.java]: Done

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestLockRemover.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+              .put(DUMMY_NAME + "." + TASK_1 + "." + FAILED_STREAM_COUNTER, 0L)
+              .build());
+    } finally {
+      closeJobClient(jobClient);
+    }
+  }
+
+  /**
+   * The test checks if the recovery watermark is only removed if the watermark has arrived from
+   * both downstream sources.
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestLockRemover.java]: Done

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 1 above)
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:115]: The assumption is:
- Every `TaskResult` is followed by a `Watermark` for normal `Trigger`s
- For the recovery `Trigger` there is no `processElement`

There are 3 possibilities:
- Normal execution - we receive a `TaskResult` and then a `Watermark` - unlocking the `lock` is handled by the `processElement`
- Recovery without ongoing execution (unlocking the `recoveryLock`) - we receive a `Watermark` without any `TaskResults` - unlocking the `recoveryLock` and a possible `lock` is handled by the `processWatermark` (the lastProcessed is `0`)
- Recovery with an ongoing execution - we receive a `TaskResult` and then a `Watermark` - unlocking the `lock` is handled by the `processElement`, unlocking the `recoveryLock` is handled by the `processWatermark` (the lastProcessed is the start time of the old task)

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResult.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+/** The result of a single Maintenance Task. */
+@Internal
+public class TaskResult {
+  private final int taskIndex;
+  private final long startEpoch;
+  private final boolean success;
+  private final List<Exception> exceptions;
+
+  public TaskResult(int taskIndex, long startEpoch, boolean success, List<Exception> exceptions) {
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResult.java:33]: should we include `taskName` so that logging can be more informative?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: > I prefer to have an output of the stream which could be logged or processed later. Having Void and changing later is backward incompatible. Having something for which we could add attributes could be compatible. That is why I have created the MaintenanceResult

unless we have a clear goal of how to use the `MaintenanceResult`. Emitting it seems a bit pre-mature to me. I don't quite get the backward incompatible part. These classes like `LockRemover` is not user-facing classes.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:115]: can you add the above context as code comment here and in the processElement method?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Done

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 1 above)
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Added the comment to the class

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+  }
+
+  @Override
+  public void processElement(StreamRecord<TaskResult> element) {
+    TaskResult taskResult = element.getValue();
+    LOG.debug("Processing task result: {}", taskResult);
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Changed

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResult.java (hunk 13 above)
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResult.java:33]: I would vote against it - changed the log to add the taskName, but I would not make it travel around the whole flow, just for logging purposes

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+/** Manages locks and collect {@link org.apache.flink.metrics.Metric} for the Maintenance Tasks. */
+@Internal
+public class LockRemover extends AbstractStreamOperator<MaintenanceResult>
+    implements OneInputStreamOperator<TaskResult, MaintenanceResult> {
+  private static final Logger LOG = LoggerFactory.getLogger(LockRemover.class);
+
+  private final TriggerLockFactory lockFactory;
+  private final List<String> taskNames;
+
+  private transient List<Counter> succeededTaskResultCounterMap;
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Done

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  private transient List<Counter> failedTaskResultCounterMap;
+  private transient List<AtomicLong> lastRunDuration;
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Because we need a way for the metrics infrastructure to get the result. Notice:
```
      getRuntimeContext()
          .getMetricGroup()
          .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
          .gauge(TableMaintenanceMetrics.LAST_RUN_DURATION_MS, length::get);
```

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    output.collect(
+        new StreamRecord<>(
+            new MaintenanceResult(
+                taskResult.startEpoch(),
+                taskResult.taskIndex(),
+                duration,
+                taskResult.success(),
+                taskResult.exceptions())));
+    lock.unlock();
+    this.lastProcessed = taskResult.startEpoch();
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Renamed

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TableMaintenanceMetrics.java [hunk 18]
```diff
@@ -28,6 +28,11 @@ public class TableMaintenanceMetrics {
   public static final String TRIGGERED = "triggered";
   public static final String NOTHING_TO_TRIGGER = "nothingToTrigger";
 
+  // LockRemover metrics
+  public static final String SUCCEEDED_TASK_COUNTER = "succeededTask";
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TableMaintenanceMetrics.java]: Changed

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    long duration = System.currentTimeMillis() - taskResult.startEpoch();
+    output.collect(
+        new StreamRecord<>(
+            new MaintenanceResult(
+                taskResult.startEpoch(),
+                taskResult.taskIndex(),
+                duration,
+                taskResult.success(),
+                taskResult.exceptions())));
+    lock.unlock();
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java:133]: You can check the Locking paragraph in the design doc: https://docs.google.com/document/d/16g3vR18mVBy8jbFaLjf2JwAANuYOmIwr15yDDxovdnA

If you still have questions, feel free to ask

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Removed

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+          getRuntimeContext()
+              .getMetricGroup()
+              .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+              .counter(TableMaintenanceMetrics.SUCCEEDED_TASK_COUNTER));
+      failedTaskResultCounters.add(
+          getRuntimeContext()
+              .getMetricGroup()
+              .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+              .counter(TableMaintenanceMetrics.FAILED_TASK_COUNTER));
+      AtomicLong length = new AtomicLong(0);
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java]: Renamed

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MaintenanceResult.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+
+/** The output of the Maintenance Flow. */
+public class MaintenanceResult {
+  private final long startEpoch;
+  private final int taskIndex;
+  private final long length;
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MaintenanceResult.java]: removed

**Comment by pvary** [comment]
@stevenzwu: I didn't get answer for my question on the Flink dev list about the output of the PostCommitTopology. See: https://lists.apache.org/thread/28qb1q6b7kz30dqjdbsw855osx4t38s7

I created some unit test and checked that if we return `Void` the PostCommitTopology is executed.

After some back-and-forth myself, I decided to move forward with the `Void` output as suggested by you. Mainly because of the `LockRemover` will be used in 2 cases:
1. `Post Commit Maintenance` - in this case the output could not be used
2. `Separate Maintenance Job` - in this case the output could be used to act upon failures, but it is unlikely and could be implemented if required later

The `LockRemover` will not be used when `External scheduler` is used.

Please share your thoughts, and review.
Thanks,
Peter

**Review by stevenzwu — APPROVED** [review]

**Comment by pvary** [comment]
Merged to main. 
Thanks for the review @stevenzwu and @advancedxy!
